internal/entitycache/cache.go

```diff
@@ -0,0 +1,669 @@
+// found, it returns an error with a params.ErrNotFound cause.
+// The returned entity will have at least the given fields filled out.
+func (c *Cache) Entity(id *charm.URL, fields ...string) (*mongodoc.Entity, error) {
+	// Start the base entity fetch asynchronously.
+	// Note that it's valid to do this because we can always tell
+	// the base entity from any URL that we can look up.
+	c.baseEntities.mu.Lock()
+	c.baseEntities.startFetch(baseURL(id))
+	c.baseEntities.mu.Unlock()
+	e, err := c.entities.entity(id, fields...)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	return e.(*mongodoc.Entity), nil
+}
+
+// BaseEntity returns the base entity with the given id. If the entity is not
+// found, it returns an error with a params.ErrNotFound cause.
+// The returned entity will have at least the given fields filled out.
+func (c *Cache) BaseEntity(id *charm.URL, fields ...string) (*mongodoc.BaseEntity, error) {
+	e, err := c.baseEntities.entity(baseURL(id), fields...)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	return e.(*mongodoc.BaseEntity), nil
+}
+
+// getEntity is used by c.entities to fetch entities.
+func (c *Cache) getEntity(id *charm.URL, fields map[string]int) (stashEntity, error) {
+	e, err := c.store.FindBestEntity(id, fields)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Any)
+	}
+	return e, nil
+}
+
+// getBaseEntity is used by c.baseEntities to fetch entities.
+func (c *Cache) getBaseEntity(id *charm.URL, fields map[string]int) (stashEntity, error) {
+	e, err := c.store.FindBaseEntity(id, fields)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Any)
+	}
+	return e, nil
+}
+
+// stash holds a set of one kind of entity (either entities or base entities).
+type stash struct {
+	// get fetches the entity with the given URL.
+	get func(id *charm.URL, fields map[string]int) (stashEntity, error)
+
+	// wg represents the set of running goroutines.
+	wg *sync.WaitGroup
+
+	// mu guards the fields below
+	mu sync.Mutex
+
+	// changed is signalled every time the entities map has changed.
+	// This means that each waiter can potentially be woken up many
+	// times before it finds the entity that it's waiting for, but
+	// saves us having a channel or condition per entity.
+	//
+	// Note that in the usual pattern we expect to see, callers
+	// will ask for entities in the same order that they arrive
+	// in the cache, so won't iterate many times.
+	changed sync.Cond
+
+	// entities holds at least one entry for each cached entity,
+	// keyed by the entity id string. A given entity always has an
+	// entry with its canonical URL as key, but also may have other
+	// entries for other unambiguous names.
+	//
+	// A nil entry indicates that the entity has been scheduled to
+	// be fetched. Entries that have been fetched but that were not
+	// found are indicated with a notFoundEntity value.
+	//
+	// Note that if an entity is in the entities stash, it does
```

Comment:
This comment seems misleading, My reading of the code so far has it that there will never be an entity and base entity in the same stash.